File: ModernJavaInAction-scala/src/modernjavainaction/chap20/Currying.scala
package modernjavainaction.chap20

object Currying {

  def main(args: Array[String]) {
    def multiply(x : Int, y: Int) = x * y
    val r1 = multiply(2, 10)
    println(r1)

    def multiplyCurry(x :Int)(y : Int) = x * y
    val r2 = multiplyCurry(2)(10)
    println(r2)

    val multiplyByTwo : Int => Int = multiplyCurry(2)
    val r3 = multiplyByTwo(10)
    println(r3)
  }

}

File: ModernJavaInAction-master/src/main/java/modernjavainaction/chap20/Currying.java
package modernjavainaction.chap20;

import java.util.function.Function;
import java.util.stream.Stream;

public class Currying {

  public static void main(String[] args) {
    int r = multiply(2, 10);
    System.out.println(r);

    Stream.of(1, 3, 5, 7)
        .map(multiplyCurry(2))
        .forEach(System.out::println);
  }

  static int multiply(int x, int y) {
    return x * y;
  }

  static Function<Integer, Integer> multiplyCurry(int x) {
    return (Integer y) -> x * y;
  }

}
